File: apps/ios/App/APIClient.swift
import Foundation
import MuralCore

final class NoRedirect: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) { completionHandler(nil) }
}

struct APIUsage { var input = 0; var output = 0; var searches = 0 }
struct APIResult { var text: String; var sources: [SourceLink]; var usage: APIUsage }

enum EndpointProtocol: String, Codable, Hashable, CaseIterable { case chatCompletions, responses }

/// A learner-owned OpenAI-compatible server. Settings live in UserDefaults; its key lives in the Keychain.
/// Voice runs turn by turn through the server's audio endpoints.
struct CustomEndpoint: Codable, Equatable {
    var enabled = false
    var baseURL = ""
    var style = EndpointProtocol.chatCompletions
    var model = ""
    var transcriptionModel = ""
    var speechModel = ""
    var voice = ""
    /// Asks reasoning models (e.g. Qwen on vLLM) to answer directly. Optional so settings saved before it still decode.
    var skipThinking: Bool?

    var url: URL? { CustomEndpointURL.parse(baseURL) }
    var textReady: Bool { url != nil && !model.isEmpty }
    var voiceReady: Bool { textReady && !transcriptionModel.isEmpty && !speechModel.isEmpty && !voice.isEmpty }

    private static let defaultsKey = "mural.customEndpoint"
    static func load() -> CustomEndpoint {
        UserDefaults.standard.data(forKey: defaultsKey).flatMap { try? JSONDecoder().decode(CustomEndpoint.self, from: $0) } ?? CustomEndpoint()
    }
    /// The endpoint personal requests use, or nil for OpenAI.
    static var active: CustomEndpoint? { let endpoint = load(); return endpoint.enabled && endpoint.textReady ? endpoint : nil }
    func save() throws { UserDefaults.standard.set(try JSONEncoder().encode(self), forKey: Self.defaultsKey) }
    static func delete() throws {
        try CredentialStore.delete(service: CredentialStore.customEndpoint)
        UserDefaults.standard.removeObject(forKey: defaultsKey)
    }
}

@MainActor final class APIClient {
    private let session: URLSession
    private struct Target { var base: URL; var key: String?; var endpoint: CustomEndpoint? }
    init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 45; config.timeoutIntervalForResource = 60
        config.httpCookieStorage = nil; config.urlCache = nil
        session = URLSession(configuration: config, delegate: NoRedirect(), delegateQueue: nil)
    }
    private func resolveTarget() throws -> Target {
        // A custom server may need no key; OpenAI always does.
        if let endpoint = CustomEndpoint.active, let base = endpoint.url {
            return Target(base: base, key: CredentialStore.read(service: CredentialStore.customEndpoint), endpoint: endpoint)
        }
        guard let key = CredentialStore.read() else { throw APIError.missingKey }
        return Target(base: URL(string: "https://api.openai.com/v1/")!, key: key, endpoint: nil)
    }
    func post(_ path: String, body: [String: Any]) async throws -> [String: Any] {
        try await postJSON(try resolveTarget(), path, body: body)
    }
    private func postJSON(_ target: Target, _ path: String, body: [String: Any]) async throws -> [String: Any] {
        let data = try await send(target, path, body: try JSONSerialization.data(withJSONObject: body), contentType: "application/json")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw APIError.invalidResponse }
        return json
    }
    /// The same limits as Android: a faulty server can't exhaust memory with an endless JSON or audio body.
    static let jsonLimit = 1_048_576, audioLimit = 16_777_216, errorBodyLimit = 16_384
    private func send(_ target: Target, _ path: String, body: Data, contentType: String, limit: Int = APIClient.jsonLimit) async throws -> Data {
        guard let url = URL(string: path, relativeTo: target.base)?.absoluteURL else { throw APIError.invalidResponse }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        if let key = target.key { request.setValue("Bearer " + key, forHTTPHeaderField: "Authorization") }
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        let (bytes, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        let succeeded = (200..<300).contains(http.statusCode)
        // Error bodies only supply a short error code, so they're read no further than that.
        let cap = succeeded ? limit : Self.errorBodyLimit
        if succeeded && http.expectedContentLength > Int64(cap) { throw APIError.invalidResponse }
        var data = Data()
        for try await byte in bytes {
            guard data.count < cap else {
                if succeeded { throw APIError.invalidResponse }
                break
            }
            data.append(byte)
        }
        try Task.checkCancellation()
        guard succeeded else {
            // OpenAI failures name OpenAI and its billing; a custom server's failure must not.
            if target.endpoint != nil { throw APIError.endpoint(http.statusCode) }
            throw ProviderFailure(status: http.statusCode, body: data, reference: http.value(forHTTPHeaderField: "x-request-id"))
        }
        return data
    }
    /// `language` is the learning language. Left to auto-detection, Whisper can turn accented Spanish into an
    /// English translation; the cost is that a reply in another language may transcribe poorly.
    func transcribe(wav: Data, language: String) async throws -> String {
        let target = try resolveTarget()
        guard let endpoint = target.endpoint else { throw APIError.invalidResponse }
        let boundary = "mural-" + UUID().uuidString
        var body = Data()
        // Whisper names Norwegian "no".
        for (name, value) in [("model", endpoint.transcriptionModel), ("response_format", "json"), ("language", language == "nb" ? "no" : language)] {
            body.append(Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n".utf8))
        }
        body.append(Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"speech.wav\"\r\nContent-Type: audio/wav\r\n\r\n".utf8))
        body.append(wav)
        body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        let data = try await send(target, "audio/transcriptions", body: body, contentType: "multipart/form-data; boundary=\(boundary)")
        guard let text = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["text"] as? String else { throw APIError.invalidResponse }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    /// Returns WAV bytes; 16-bit PCM WAV is the one format every compatible speech server offers.
    func speak(_ text: String) async throws -> Data {
        let target = try resolveTarget()
        guard let endpoint = target.endpoint else { throw APIError.invalidResponse }
        let body: [String: Any] = ["model": endpoint.speechModel, "voice": endpoint.voice, "input": text, "response_format": "wav"]
        return try await send(target, "audio/speech", body: try JSONSerialization.data(withJSONObject: body), contentType: "application/json", limit: Self.audioLimit)
    }
    func respond(instructions: String, input: String, schema: [String: Any]? = nil, search: Bool = false, onText: (@MainActor (String) -> Void)? = nil) async throws -> APIResult {
        let target = try resolveTarget()
        if let endpoint = target.endpoint, endpoint.style == .chatCompletions {
            // No standard web search exists in Chat Completions, so `search` is ignored and topics stay unsourced.
            var body: [String: Any] = ["model": endpoint.model, "messages": [["role": "system", "content": instructions], ["role": "user", "content": input]]]
            // vLLM's chat-template switch; OpenAI and some other servers reject unknown fields, so it's opt-in.
            if endpoint.skipThinking == true { body["chat_template_kwargs"] = ["enable_thinking": false] }
            if let schema { body["response_format"] = ["type": "json_schema", "json_schema": ["name": "mural_result", "strict": true, "schema": schema]] }
            let result = try Self.decodeChatCompletion(try await postJSON(target, "chat/completions", body: body))
            onText?(result.text) // A custom endpoint answers in one piece.
            return result
        }
        var body: [String: Any] = ["model": target.endpoint?.model ?? "gpt-5.6-luna", "store": false, "instructions": instructions,
                                  "input": [["role": "user", "content": input]], "max_output_tokens": schema == nil ? 1400 : 2200,
                                  "reasoning": ["effort": "low"]]
        if let schema { body["text"] = ["format": ["type": "json_schema", "name": "mural_result", "strict": true, "schema": schema]] }
        // Web search is OpenAI's hosted tool; a compatible Responses server may reject it.
        if search && target.endpoint == nil { body["tools"] = [["type": "web_search"]]; body["tool_choice"] = "auto"; body["max_tool_calls"] = 1 }
        let json: [String: Any]
        // OpenAI streams meanings as they arrive; a custom Responses server answers in one piece.
        if let onText, target.endpoint == nil {
            body["stream"] = true
            json = try await streamResponse(target, body: body, onText: onText)
        } else { json = try await postJSON(target, "responses", body: body) }
        guard json["status"] as? String == "completed" else { throw APIError.incomplete }
        var text = "", sources: [SourceLink] = [], usage = APIUsage()
        for item in json["output"] as? [[String: Any]] ?? [] {
            if item["type"] as? String == "web_search_call" { usage.searches += 1 }
            for content in item["content"] as? [[String: Any]] ?? [] {
                if content["type"] as? String == "refusal" { throw APIError.refused }
                if content["type"] as? String == "output_text" { text += content["text"] as? String ?? "" }
                for citation in content["annotations"] as? [[String: Any]] ?? [] {
                    guard citation["type"] as? String == "url_citation", let url = citation["url"] as? String else { continue }
                    let source = SourceLink(title: citation["title"] as? String ?? "Source", url: url)
                    if source.safeURL != nil && !sources.contains(where: { $0.url == url }) { sources.append(source) }
                }
            }
        }
        if let u = json["usage"] as? [String: Any] { usage.input = u["input_tokens"] as? Int ?? 0; usage.output = u["output_tokens"] as? Int ?? 0 }
        // A local reasoning model behind a Responses endpoint can inline its thinking too.
        text = text.replacingOccurrences(of: Self.thinking, with: "", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { throw APIError.incomplete }
        if target.endpoint != nil { onText?(text) }
        return APIResult(text: text, sources: sources, usage: usage)
    }
    /// Inline reasoning, including a block the server cut off before its closing tag.
    static let thinking = "<think>[\\s\\S]*?(?:</think>|$)"
    /// Compatible servers carry no web citations. Local reasoning models may inline thinking, which is never spoken or stored.
    static func decodeChatCompletion(_ json: [String: Any]) throws -> APIResult {
        guard let choice = (json["choices"] as? [[String: Any]])?.first, let message = choice["message"] as? [String: Any] else { throw APIError.incomplete }
        let finish = choice["finish_reason"] as? String
        if message["refusal"] is String || finish == "content_filter" { throw APIError.refused }
        if finish == "length" { throw APIError.incomplete }
        let text = (message["content"] as? String ?? "")
            .replacingOccurrences(of: Self.thinking, with: "", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { throw APIError.incomplete }
        let usage = json["usage"] as? [String: Any]
        return APIResult(text: text, sources: [], usage: APIUsage(input: usage?["prompt_tokens"] as? Int ?? 0, output: usage?["completion_tokens"] as? Int ?? 0))
    }
    private func streamResponse(_ target: Target, body: [String: Any], onText: @MainActor (String) -> Void) async throws -> [String: Any] {
        guard let key = target.key, let url = URL(string: "responses", relativeTo: target.base)?.absoluteURL else { throw APIError.missingKey }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer " + key, forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (bytes, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            var body = Data()
            for try await byte in bytes { body.append(byte); if body.count >= 16_384 { break } }
            throw ProviderFailure(status: http.statusCode, body: body, reference: http.value(forHTTPHeaderField: "x-request-id"))
        }
        guard http.value(forHTTPHeaderField: "Content-Type")?.lowercased().hasPrefix("text/event-stream") == true else { throw APIError.invalidResponse }
        var decoder = ResponseTextStream()
        do {
            for try await byte in bytes {
                try Task.checkCancellation()
                guard let update = try decoder.consume(byte: byte) else { continue }
                switch update {
                case .text(let value): onText(value)
                case .completed(let result): return result
                }
            }
        } catch ResponseTextStream.Failure.refused { throw APIError.refused }
        catch is ResponseTextStream.Failure { throw APIError.incomplete }
        throw APIError.incomplete
    }
    static func object(_ fields: [String: Any]) -> [String: Any] { ["type": "object", "properties": fields, "required": fields.keys.sorted(), "additionalProperties": false] }
    static let string: [String: Any] = ["type": "string"]
    static func assessmentSchema(language: LanguageModule) -> [String: Any] { object([
        "outcome": ["type": "string", "enum": ["success", "partial", "breakdown", "uncertain"]],
        "suggestedLevel": ["type": "integer", "minimum": 0, "maximum": 5], "nextGoal": string, "capability": string,
        "words": ["type": "array", "maxItems": 12, "items": object([
            "lemma": string, "meaning": string, "form": string, "quote": string, "language": ["type": "string", "enum": Array(Set([language.id, "en", "mixed", "uncertain"])).sorted()],
            "kind": ["type": "string", "enum": ["exposure", "understanding", "assisted", "independent", "lapse"]],
            "confidence": ["type": "number", "minimum": 0, "maximum": 1], "sourceIDs": ["type": "array", "items": string]
        ])]
    ]) }
    enum APIError: LocalizedError {
        case missingKey, invalidResponse, incomplete, refused, http(Int), endpoint(Int)
        var errorDescription: String? {
            switch self {
            case .endpoint(let status): "Your custom endpoint couldn’t complete the request (HTTP \(status)). Check its settings."
            case .missingKey: "Add your OpenAI key in Settings to begin."
            case .invalidResponse, .incomplete: CustomEndpoint.active == nil
                ? "OpenAI returned an incomplete response. Please try again."
                : "Your custom endpoint returned an incomplete or unexpected response. Check its API style and models."
            case .refused: "Mural couldn’t complete that request. Try a different topic."
            case .http(401): "Your OpenAI key wasn’t accepted. Check it in Settings."
            case .http(403), .http(404): "This API key may not have access to the requested model. Check your OpenAI project."
            case .http(429): "OpenAI’s usage or rate limit was reached. Check your project’s billing and limits."
            case .http(let status): "OpenAI couldn’t complete the request (HTTP \(status)). Please try again."
            }
        }
    }
}

File: apps/android/app/src/main/java/chat/mural/network/APIClient.kt
package chat.mural.network

import chat.mural.core.SourceLink
import chat.mural.core.ProviderFailureKind
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.Call
import okhttp3.Callback
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okio.Buffer

data class APIUsage(val input: Int = 0, val output: Int = 0, val searches: Int = 0)
data class APIResult(val text: String, val sources: List<SourceLink>, val usage: APIUsage)

class APIClient internal constructor(
    private val target: () -> EndpointTarget,
    private val client: OkHttpClient,
) : TeachingClient, LiveSessionProvider {
    /** Resolves OpenAI or the learner's custom endpoint for each request. */
    constructor(target: () -> EndpointTarget) : this(target, defaultClient())

    internal constructor(key: String?, client: OkHttpClient, baseUrl: HttpUrl) :
        this({ openAI(key, baseUrl) }, client)

    override suspend fun createLiveSession(request: LiveSessionRequest): LiveSessionConnection {
        val result = post("live/sessions", buildJsonObject {
            put("session", buildJsonObject {
                put("model", "gpt-live-1"); put("instructions", request.instructions); put("input", request.history)
                put("store", false)
                put("delegation", buildJsonObject { put("type", "client") })
                put("audio", buildJsonObject { put("output", buildJsonObject { put("voice", "marin") }) })
            })
            put("transport", buildJsonObject { put("type", "webrtc"); put("sdp", request.sdp) })
        })
        val transport = result["transport"] as? JsonObject ?: throw APIException.InvalidResponse
        val answer = (transport["sdp"] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
        if (transport["type"] != JsonPrimitive("webrtc") || answer.isNullOrBlank()) throw APIException.InvalidResponse
        val id = ((result["session"] as? JsonObject)?.get("id") as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
        return LiveSessionConnection(answer, id)
    }

    suspend fun post(path: String, body: JsonObject): JsonObject = postJson(target(), path, body)

    private suspend fun postJson(endpoint: EndpointTarget, path: String, body: JsonObject): JsonObject =
        execute(endpoint, path, body.toString().toRequestBody(JSON_MEDIA_TYPE), MAX_RESPONSE_BYTES) { payload ->
            try { JSON.parseToJsonElement(payload.toString(Charsets.UTF_8)).jsonObject }
            catch (_: Exception) { throw APIException.InvalidResponse }
        }

    private suspend fun <T> execute(endpoint: EndpointTarget, path: String, body: RequestBody, limit: Long, parse: (ByteArray) -> T): T {
        if (!VALID_PATH.matches(path) || path.contains("..") || path.startsWith('/')) {
            throw APIException.InvalidResponse
        }
        // A custom server may need no key; OpenAI always does.
        val key = endpoint.key ?: if (endpoint.custom) null else throw APIException.MissingKey
        val request = Request.Builder()
            .url(endpoint.baseUrl.newBuilder().addPathSegments(path).build())
            .apply { if (key != null) header("Authorization", "Bearer $key") }
            .post(body)
            .build()

        // Parse on OkHttp's worker while the continuation remains cancellable.
        // Cancellation closes a response even if the peer stalls halfway through its body.
        return suspendCancellableCoroutine { continuation ->
            val call = client.newCall(request)
            continuation.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onFailure(call: Call, error: IOException) {
                    if (continuation.isActive) continuation.resumeWithException(error)
                }
                override fun onResponse(call: Call, response: Response) {
                    try {
                        val value = response.use {
                            if (it.code !in 200..299) {
                                val errorCode = runCatching {
                                    val payload = it.peekBody(16_385).string()
                                    if (payload.toByteArray(Charsets.UTF_8).size > 16_384) null else
                                        (JSON.parseToJsonElement(payload).jsonObject["error"] as? JsonObject)
                                            ?.get("code")?.jsonPrimitive?.contentOrNull
                                }.getOrNull()
                                throw APIException.Http(it.code, errorCode, it.header("x-request-id"))
                            }
                            parse(it.readBoundedBody(limit))
                        }
                        if (continuation.isActive) continuation.resume(value)
                    } catch (error: Exception) {
                        if (continuation.isActive) continuation.resumeWithException(error)
                    }
                }
            })
        }
    }

    override suspend fun respond(
        instructions: String,
        input: String,
        schema: JsonObject?,
        search: Boolean,
        purpose: HelperPurpose?,
    ): APIResult {
        val endpoint = target()
        return if (endpoint.protocol == EndpointProtocol.CHAT_COMPLETIONS)
            decodeChatCompletion(postJson(endpoint, "chat/completions", chatBody(endpoint, instructions, input, schema)))
        // Web search is OpenAI's hosted tool; a compatible Responses server may reject it.
        else decodeTeachingResponse(postJson(endpoint, "responses", responsesBody(endpoint.model, instructions, input, schema, search && !endpoint.custom)))
    }

    // ponytail: Chat Completions has no standard web search, so `search` is ignored and topics stay unsourced.
    private fun chatBody(endpoint: EndpointTarget, instructions: String, input: String, schema: JsonObject?) = buildJsonObject {
        put("model", endpoint.model)
        // vLLM's chat-template switch; OpenAI and some other servers reject unknown fields, so it's opt-in.
        if (endpoint.skipThinking) put("chat_template_kwargs", buildJsonObject { put("enable_thinking", false) })
        put("messages", buildJsonArray {
            add(buildJsonObject { put("role", "system"); put("content", instructions) })
            add(buildJsonObject { put("role", "user"); put("content", input) })
        })
        if (schema != null) {
            put("response_format", buildJsonObject {
                put("type", "json_schema")
                put("json_schema", buildJsonObject { put("name", "mural_result"); put("strict", true); put("schema", schema) })
            })
        }
    }

    private fun responsesBody(model: String, instructions: String, input: String, schema: JsonObject?, search: Boolean) =
        buildJsonObject {
            put("model", model)
            put("store", false)
            put("instructions", instructions)
            put("input", buildJsonArray {
                add(buildJsonObject {
                    put("role", "user")
                    put("content", input)
                })
            })
            put("max_output_tokens", if (schema == null) 1_400 else 2_200)
            put("reasoning", buildJsonObject { put("effort", "low") })
            if (schema != null) {
                put("text", buildJsonObject {
                    put("format", buildJsonObject {
                        put("type", "json_schema")
                        put("name", "mural_result")
                        put("strict", true)
                        put("schema", schema)
                    })
                })
            }
            if (search) {
                put("tools", buildJsonArray { add(buildJsonObject { put("type", "web_search") }) })
                put("tool_choice", "auto")
                put("max_tool_calls", 1)
            }
        }

    /** Streams OpenAI meanings as they arrive. A custom endpoint answers in one piece, since compatible servers
     * don't reliably emit OpenAI's Responses stream events. */
    override suspend fun streamMeaning(instructions: String, input: String, onText: (String) -> Unit): APIResult {
        val endpoint = target()
        if (endpoint.custom) return respond(instructions, input, purpose = HelperPurpose.MEANING).also { onText(it.text) }
        val key = endpoint.key ?: throw APIException.MissingKey
        val body = buildJsonObject {
            responsesBody(endpoint.model, instructions, input, null, false).forEach { (key, value) -> put(key, value) }
            put("stream", true)
        }
        val request = Request.Builder().url(endpoint.baseUrl.newBuilder().addPathSegments("responses").build())
            .header("Authorization", "Bearer $key").header("Accept", "text/event-stream")
            .post(body.toString().toRequestBody(JSON_MEDIA_TYPE)).build()
        val callbacks = currentCoroutineContext().minusKey(Job)
        val result = streamingResponse(client, request) { response ->
            if (!response.isSuccessful) {
                val code = runCatching { JSON.parseToJsonElement(response.peekBody(16_384).string()).jsonObject["error"]
                    ?.jsonObject?.get("code")?.jsonPrimitive?.contentOrNull }.getOrNull()
                throw APIException.Http(response.code, code, response.header("x-request-id"))
            }
            if (response.header("Content-Type")?.startsWith("text/event-stream", ignoreCase = true) != true) throw APIException.InvalidResponse
            var text = ""
            readTextEvents(response) { event ->
                when (event["type"]?.jsonPrimitive?.contentOrNull) {
                    "response.output_text.delta" -> {
                        text += event["delta"]?.jsonPrimitive?.contentOrNull ?: throw APIException.InvalidResponse
                        if (text.toByteArray(Charsets.UTF_8).size > 65_536) throw APIException.InvalidResponse
                        withContext(callbacks) { onText(text) }; null
                    }
                    "response.completed" -> event["response"] as? JsonObject ?: throw APIException.Incomplete
                    "response.refusal.delta", "response.refusal.done" -> throw APIException.Refused
                    "error", "response.failed", "response.incomplete" -> throw APIException.Incomplete
                    else -> null
                }
            }
        }
        return decodeTeachingResponse(result)
    }

    /** [language] is the learning language. Left to auto-detection, Whisper can turn accented Spanish into an
     * English translation; the cost is that a reply in another language may transcribe poorly. */
    suspend fun transcribe(wav: ByteArray, language: String): String {
        val endpoint = target()
        val body = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("model", endpoint.transcriptionModel)
            .addFormDataPart("response_format", "json")
            .addFormDataPart("language", if (language == "nb") "no" else language) // Whisper names Norwegian "no".
            .addFormDataPart("file", "speech.wav", wav.toRequestBody(WAV_MEDIA_TYPE))
            .build()
        return execute(endpoint, "audio/transcriptions", body, MAX_RESPONSE_BYTES) { payload ->
            val json = try { JSON.parseToJsonElement(payload.toString(Charsets.UTF_8)).jsonObject }
                catch (_: Exception) { throw APIException.InvalidResponse }
            (json["text"] as? JsonPrimitive)?.takeIf { it.isString }?.content?.trim() ?: throw APIException.InvalidResponse
        }
    }

    /** Returns WAV bytes; 16-bit PCM WAV is the one format every compatible speech server offers. */
    suspend fun speak(text: String): ByteArray {
        val endpoint = target()
        val body = buildJsonObject {
            put("model", endpoint.speechModel); put("voice", endpoint.voice); put("input", text); put("response_format", "wav")
        }
        return execute(endpoint, "audio/speech", body.toString().toRequestBody(JSON_MEDIA_TYPE), MAX_AUDIO_BYTES) { it }
    }

    private fun Response.readBoundedBody(limit: Long): ByteArray {
        val responseBody = body ?: throw APIException.InvalidResponse
        if (responseBody.contentLength() > limit) throw APIException.InvalidResponse
        val source = responseBody.source()
        val buffer = Buffer()
        var total = 0L
        while (true) {
            val count = source.read(buffer, minOf(8_192L, limit + 1L - total))
            if (count == -1L) break
            total += count
            if (total > limit) throw APIException.InvalidResponse
        }
        return buffer.readByteArray()
    }

    sealed class APIException(message: String, cause: Throwable? = null) : IOException(message, cause) {
        data object MissingKey : APIException("Add your OpenAI key in Settings to begin.")
        data object InvalidResponse : APIException("OpenAI returned an incomplete response. Please try again.")
        data object Incomplete : APIException("OpenAI returned an incomplete response. Please try again.")
        data object Refused : APIException("Mural couldn't complete that request. Try a different topic.")
        class Http(val status: Int, code: String? = null, reference: String? = null) : APIException(messageFor(status)) {
            val code = ProviderFailureKind.safeCode(code)
            val reference = ProviderFailureKind.safeReference(reference)
            val kind get() = ProviderFailureKind.classify(status, code)
        }

        companion object {
            private fun messageFor(status: Int): String = when (status) {
                401 -> "Your OpenAI key wasn't accepted. Check it in Settings."
                403, 404 -> "This API key may not have access to the requested model. Check your OpenAI project."
                429 -> "OpenAI's usage or rate limit was reached. Check your project billing and limits."
                else -> "OpenAI couldn't complete the request (HTTP $status). Please try again."
            }
        }
    }

    companion object {
        const val TEACHER_MODEL = "gpt-5.6-luna"
        private val API_BASE_URL = HttpUrl.Builder()
            .scheme("https")
            .host("api.openai.com")
            .addPathSegment("v1")
            .addPathSegment("")
            .build()
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        private val WAV_MEDIA_TYPE = "audio/wav".toMediaType()
        private val VALID_PATH = Regex("[a-z0-9][a-z0-9_/-]*")
        private const val MAX_RESPONSE_BYTES = 1_048_576L
        private const val MAX_AUDIO_BYTES = 16_777_216L
        private val JSON = Json { ignoreUnknownKeys = true }

        fun openAI(key: String?, baseUrl: HttpUrl = API_BASE_URL) =
            EndpointTarget(baseUrl, key, EndpointProtocol.RESPONSES, TEACHER_MODEL)

        private fun defaultClient() = OkHttpClient.Builder()
            .connectTimeout(45, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(45, TimeUnit.SECONDS)
            .callTimeout(60, TimeUnit.SECONDS)
            .followRedirects(false)
            .followSslRedirects(false)
            .cookieJar(CookieJar.NO_COOKIES)
            .cache(null)
            .build()


    }
}
